### PR TITLE
AdvancedDigitizing add Z/M support

### PR DESCRIPTION
**Pull request by lbartoletti**
## Description

Drawing directly vertices with a Z other than the default one is not possible today in QGIS. We proceed as follows:

advanced_zm_before mkv

There have been several requests to add the Z and M in this panel. At least by @saberraz and @DelazJ in https://github.com/qgis/QGIS/issues/25865 and https://github.com/qgis/QGIS/issues/29415 

Some people need more and more to draw networks with 3D vertices : it is particularly interesting to propose the addition of the Z (and M) in this one.

I simply propose the addition of these two lines in the panel to remain in the same logic as today:

advanced_zm_after mkv

There have been several developments to fix Z snapping on vertices and segments (but, for this one, only when topological editing is activated) and this PR will not break this logic. The weight of the snapping is higher than the value given in the widget.

advanced_zm_snap mkv

About the code, it is mainly a copy and paste of existing functions for X/Y with adaptation to get the Z (and M)

Still in WIP state since I have to continue some tests. Feedbacks are appreciated.

Sponsored by: Métropole Européenne de Lille @Jean-Roc

### CODE REVIEW

**Comment by nyalldawson** [comment]
> There have been several developments to fix Z snapping on vertices and segments (but, for this one, only when topological editing is activated) and this PR will not break this logic. The weight of the snapping is higher than the value given in the widget.

I'm unsure about this -- if I'm interpreting correctly it means that I manually type a value of "15" in the dock, only to have it silently be replaced by a snapped z value of 5 (with no notification to me). I think this logic is backwards, and an explicit z/m value should take precedence. 

I can see two ways to approach this:
1 . default to null in the dock, and null means "use default or get from snapped vertex"
2. if the dock z/m constraint isn't locked then it updates in realtime to show the snapped z/m value

What's your thoughts?

**Comment by lbartoletti** [comment]
> I'm unsure about this -- if I'm interpreting correctly it means that I manually type a value of "15" in the dock, only to have it silently be replaced by a snapped z value of 5 (with no notification to me). I think this logic is backwards, and an explicit z/m value should take precedence.

Yes, you're right. As mentioned, this PR is still in a WIP state and more particularly on this topic. As it stands, snapping overrides the value in the dock. I started to think about what to prioritize (dock or snapping) and preferred to open this PR to collect feedback. So, thanks for your feedback.

> 1 . default to null in the dock, and null means "use default or get from snapped vertex"
> 2. if the dock z/m constraint isn't locked then it updates in realtime to show the snapped z/m value

At the moment, in this PR, if the dock z/m constraint isn't locked, the value is 0.0 for m and `defaultZValue()` for Z.
If a vertex is snapped it will be the Z/M value from the vertex
If a segment is snapped AND if the topology editing is active it will the Z/M interpolated.

So, if it's z/m constraint is locked and we snap a vertex, should we take the Z/M value of the vertex or the widget value? Same for the segment, when applicable.

In a way it seems coherent to respect the snapping for consistency with the existing data and the drawing process. In another way, if the user indicates a constraint, it means that he has determined it.

I don't know yet what to choose. Should we remove the constraint if there is a snap with a message warning of this?

**Comment by nyalldawson** [comment]
> I don't know yet what to choose. Should we remove the constraint if there is a snap with a message warning of this?

I think the options are:
1. Change the constraint to show "auto" by default, which means "get from snapping/global default", but if a user manually changes this then the manually entered value will be used REGARDLESS of if the point is snapped 
or
2. Dynamically update the value in the dock when it's NOT locked by the user and snapping occurs

**Comment by nyalldawson** [comment]
(i'm fine with either of those options!)

**Comment by lbartoletti** [comment]
@nyalldawson after all my PR to fix M values, I finally finish this one \o/

So, for the "conflict" between snapping and the value in the dock. I finally choose the second option:

The value in the dock is preferred IF the lock button is checked, instead we use the snapped value if applicable.


https://user-images.githubusercontent.com/7521540/132009572-53082a9b-5c57-4010-8439-2bef62afeea5.mp4

**Comment by nyalldawson** [comment]
@lbartoletti somehow i missed your update here! anyway, the new behavior looks great to me.

one question -- is this still subject to the "only projected crs" constraint?

**Comment by lbartoletti** [comment]
> @lbartoletti somehow i missed your update here! 

no worries :)

> anyway, the new behavior looks great to me.

thanks!

> one question -- is this still subject to the "only projected crs" constraint?

Yes, since some constraints are calculated using cartesian methods.

File: python/gui/auto_generated/qgsadvanceddigitizingdockwidget.sip.in
Comment (nyalldawson) [inline python/gui/auto_generated/qgsadvanceddigitizingdockwidget.sip.in:486]: api break

File: src/core/qgscadutils.h
Comment (nyalldawson) [inline src/core/qgscadutils.h:65]: \since QGIS 3.22
(elsewhere too)

File: src/gui/qgsadvanceddigitizingdockwidget.cpp
Comment (nyalldawson) [inline src/gui/qgsadvanceddigitizingdockwidget.cpp:323]:
```suggestion
  if (  QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( mMapCanvas->currentLayer() ) )
```

File: src/gui/qgsadvanceddigitizingdockwidget.cpp
Comment (nyalldawson) [inline src/gui/qgsadvanceddigitizingdockwidget.cpp:1222]: ctrl + z is a bad combination -- that's universally used for undo

**Comment by nyalldawson** [comment]
Looking good to me -- just the api break mostly which needs fixing. I'd approve this if not for that, so I've marked it as freeze exempt just in case you can't fix the break before today's freeze...

**Comment by nyalldawson** [comment]
@lbartoletti 

> Yes, since some constraints are calculated using cartesian methods.

Right, but that doesn't apply to z or m (or even x and y), only the distance/angle based constraints. I think we should loosen this restriction and allow just the x/y/z/m constraints to be used when cad dock is enabled.

File: python/gui/auto_generated/qgsadvanceddigitizingdockwidget.sip.in
Comment (lbartoletti) [reply python/gui/auto_generated/qgsadvanceddigitizingdockwidget.sip.in:307]: Yes it's an API break :/ but is it really a problem? Since QgsPointXY can be constructed from a QgsPoint (the reverse is not true)

Comment (nyalldawson) [reply python/gui/auto_generated/qgsadvanceddigitizingdockwidget.sip.in:307]: It is -- depending on what methods the python script plugs the result into they may get errors

Comment (lbartoletti) [reply python/gui/auto_generated/qgsadvanceddigitizingdockwidget.sip.in]: OK! Fixed. Thanks.

File: python/core/auto_generated/qgscadutils.sip.in
Comment (lbartoletti) [reply python/core/auto_generated/qgscadutils.sip.in]: Done. Thanks for the pointers!

File: python/gui/auto_generated/qgsadvanceddigitizingdockwidget.sip.in
Comment (lbartoletti) [reply python/gui/auto_generated/qgsadvanceddigitizingdockwidget.sip.in]: Done.

**Comment by lbartoletti** [comment]
> Looking good to me -- just the api break mostly which needs fixing. I'd approve this if not for that, so I've marked it as freeze exempt just in case you can't fix the break before today's freeze...

Many thanks Nyall :pray: 

> @lbartoletti
> 
> > Yes, since some constraints are calculated using cartesian methods.
> 
> Right, but that doesn't apply to z or m (or even x and y), only the distance/angle based constraints. I think we should loosen this restriction and allow just the x/y/z/m constraints to be used when cad dock is enabled.

Yes, indeed, you may want to add the geographic coordinates. It will be necessary to adjust the precision of the text in this case. 6 is too much for Cartesian (3 is more than enough), but you may want 8 digits for geographic and disable angle/distance widgets. I will prepare a dedicated PR after this one is merged.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**
